polynote-frontend/polynote/ui/component/comment.ts

```diff
@@ -0,0 +1,517 @@
+
+    constructor(readonly cellId: number, readonly editor: IStandaloneCodeEditor) {
+        super();
+
+        this.subscribe(CreateComment, (cellId, comment) => {
+            this._add(comment);
+        });
+        this.subscribe(UpdateComment, (cellId, commentId, range, content) => {
+            this._update(commentId, range, content);
+        });
+        this.subscribe(DeleteComment, (cellId, commentId) => {
+            this._delete(commentId);
+        });
+
+        this.editor.onDidChangeModelContent((evt: IModelContentChangedEvent) => {
+            const model = this.editor.getModel();
+            if (model) {
+                // check for drift between decorations and the ranges we're keeping
+                const modelDecorations = model.getAllDecorations();
+                Object.entries(this.highlights).forEach(([range_str, highlights]) => {
+                    const maybeDecoration = modelDecorations.find(d => highlights.includes(d.id));
+                    if (maybeDecoration && !maybeDecoration.range.isEmpty()) {
+                        const ps = PosRange.fromString(range_str);
+                        const startPos = model.getPositionAt(ps.start);
+                        const endPos = model.getPositionAt(ps.end);
+                        const mRange = monaco.Range.fromPositions(startPos, endPos);
+                        if (!monaco.Range.equalsRange(maybeDecoration.range, mRange)) {
+                            // we have a highlight with the same ID, but a different range. This means there is some drift.
+                            const newRange = new PosRange(model.getOffsetAt(maybeDecoration.range.getStartPosition()), model.getOffsetAt(maybeDecoration.range.getEndPosition()));
+                            const rootId = this.commentRoots[range_str];
+                            const root = this.comments[rootId];
+
+                            this.publish(new UpdateComment(this.cellId, rootId, newRange, root.content));
+
+                            // reset highlights.
+                            delete this.highlights[range_str];
+                            this.highlights[newRange.asString] = highlights;
+                        }
+                    } else {
+                        // decoration wasn't found or was empty, so we need to delete it.
+                        const rootAtRange = this.commentRoots[range_str];
+                        if (rootAtRange) {
+                            this.publish(new DeleteComment(this.cellId, rootAtRange));
+                        }
+                        delete this.highlights[range_str];
+
+                        // if the range was empty, remove it.
+                        if (maybeDecoration) model.deltaDecorations([maybeDecoration.id], [])
+                    }
+                });
+            }
+        })
+    }
+
+    // placeholder for avatars
+    private static fetchAvatar(author: string) {
```

Comment:
This placeholder is a possible entry point for whatever avatar implementation we might have in the future.